File: leetcode/javascript/15.js
// Leetcode 15 - 3Sum
// Nov. 29, 2023

/**
 * @param {number[]} sortedNums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    let sortedNums = nums.sort((a,b) => { return a - b; });
    let ans = [];
    for (let i = 0; i < sortedNums.length - 2; i++) {
        if (i > 0 && sortedNums[i] == sortedNums[i-1]) continue;
        let l = i + 1, r = sortedNums.length - 1;
        while (l < r) {
            let total = sortedNums[i] + sortedNums[l] + sortedNums[r];
            if (total === 0) {
                ans.push([sortedNums[i], sortedNums[l], sortedNums[r]]);
                while (l < r && sortedNums[l] === sortedNums[l+1]) l++;
                while (l < r && sortedNums[r] === sortedNums[r-1]) r--;
                l++;
                r--;
            } 
            else if (total < 0) l++;
            else r--;
        }
    }

    return ans;
};

File: leetcode/c++/15.cpp
// Leetcode 15 - 3Sum
// Nov. 29, 2023

#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ans;
        sort(nums.begin(), nums.end());

        for (int i = 0; i < nums.size() - 2; i++) {
            if (i == 0 || nums[i] != nums[i-1]) {
                int l = i + 1, r = nums.size() - 1;
                while (l < r) {
                    int total = nums[i] + nums[l] + nums[r];
                    if (total < 0 || (l > i + 1 && nums[l] == nums[l-1])) l++;
                    else if (total > 0 || (r < nums.size() - 1 && nums[r] == nums[r+1])) r--;
                    else ans.push_back({nums[i], nums[l++], nums[r--]});
                }
            }
        }

        return ans;
    }
};
